core/src/main/java/com/linecorp/armeria/client/endpoint/CompositeEndpointGroup.java

```diff
@@ -17,148 +17,103 @@
-                           Iterable<EndpointGroup> endpointGroups) {
-
-        this.endpointGroups = ImmutableList.copyOf(requireNonNull(endpointGroups, "endpointGroups"));
-        dirty = new AtomicBoolean(true);
+    public CompositeEndpointGroup(EndpointSelectionStrategy selectionStrategy,
+                                  Iterable<EndpointGroup> endpointGroups) {
+        super(selectionStrategy, true, selectionTimeoutMillis(endpointGroups));
+        setEndpointGroups(ImmutableSet.copyOf(endpointGroups));
+    }
 
+    static long selectionTimeoutMillis(Iterable<EndpointGroup> endpointGroups) {
         long selectionTimeoutMillis = 0;
-        for (EndpointGroup endpointGroup : endpointGroups) {
-            endpointGroup.addListener(unused -> {
-                dirty.set(true);
-                notifyListeners(endpoints());
-            });
-            selectionTimeoutMillis = Math.max(selectionTimeoutMillis,
-                                              endpointGroup.selectionTimeoutMillis());
+        for (EndpointGroup endpointGroup: endpointGroups) {
+            selectionTimeoutMillis = Math.max(selectionTimeoutMillis, endpointGroup.selectionTimeoutMillis());
         }
-        this.selectionTimeoutMillis = selectionTimeoutMillis;
-
-        initialEndpointsFuture =
-                CompletableFuture.anyOf(this.endpointGroups.stream()
-                                                           .map(EndpointGroup::whenReady)
-                                                           .toArray(CompletableFuture[]::new))
-                                 .thenApply(unused -> endpoints());
-
-        this.selectionStrategy = requireNonNull(selectionStrategy, "selectionStrategy");
-        selector = requireNonNull(selectionStrategy, "selectionStrategy").newSelector(this);
+        return selectionTimeoutMillis;
     }
 
-    @Override
-    public List<Endpoint> endpoints() {
-        if (!dirty.get()) {
-            return merged;
+    /**
+     * Sets the {@link EndpointGroup}s contained by this {@link CompositeEndpointGroup}.
+     * {@link EndpointGroup}s that aren't used anymore are closed and cleaned up automatically.
+     */
+    protected final void setEndpointGroups(Set<EndpointGroup> endpointGroups) {
+        lock.lock();
+        try {
+            final Set<EndpointGroup> toRemove = Sets.difference(this.endpointGroups, endpointGroups);
+            final Set<EndpointGroup> toAdd = Sets.difference(endpointGroups, this.endpointGroups);
+
+            this.endpointGroups = ImmutableSet.copyOf(requireNonNull(endpointGroups, "endpointGroups"));
+            toRemove.forEach(endpointGroup -> endpointGroup.removeListener(endpointsUpdatedCallback));
+            toAdd.forEach(endpointGroup -> endpointGroup.addListener(endpointsUpdatedCallback, true));
+            if (toAdd.isEmpty() && !toRemove.isEmpty()) {
+                // Callback should be invoked at least once after endpoint groups are updated
```

Comment:
`CompositeEndpointGroup.whenReady()` may return incorrect status when an `EndpointGroup` is removed before some `EndpointGroup` is ready.
```java
class CompositeEndpointGroupTest {

    @Test
    void shouldNotCompleteUntilUpdated() {
        final DynamicEndpointGroup groupA = new DynamicEndpointGroup();
        final DynamicEndpointGroup groupB = new DynamicEndpointGroup();
        final TestCompositeEndpointGroup endpointGroup = new TestCompositeEndpointGroup(
                EndpointSelectionStrategy.roundRobin(), ImmutableList.of(groupA, groupB));
        assertThat(endpointGroup.whenReady()).isNotDone();
        endpointGroup.updateEndpointGroups(ImmutableSet.of(groupA));
        assertThat(endpointGroup.whenReady()).isNotDone();
    }

    class TestCompositeEndpointGroup extends CompositeEndpointGroup {
        TestCompositeEndpointGroup(EndpointSelectionStrategy selectionStrategy, Iterable<EndpointGroup> endpointGroups) {
            super(selectionStrategy, endpointGroups);
        }

        public void updateEndpointGroups(Set<EndpointGroup> endpointGroups) {
            setEndpointGroups(endpointGroups);
        }
    }
}

```